src/main/java/io/airlift/units/ThreadsCount.java

```
@@ -0,0 +1,195 @@
+        }
+    }
+
+    static int nextPowerOfTwo(int x)
+    {
+        if (x == 0) {
+            return 1;
+        }
+
+        --x;
+        x |= x >> 1;
+        x |= x >> 2;
+        x |= x >> 4;
+        x |= x >> 8;
+        return (x | x >> 16) + 1;
+    }
+
+    @Override
+    public int compareTo(ThreadsCount o)
+    {
+        return Integer.compare(threadsCount, o.threadsCount);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        ThreadsCount that = (ThreadsCount) o;
+        return threadsCount == that.threadsCount;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return threadsCount;
+    }
+
+    @Override
+    public String toString()
+    {
+        if (threadsCount == 1) {
+            return "1 thread";
+        }
+        return format("%d threads", threadsCount);
+    }
+
+    static final class MachineInfo
+    {
+        private static final Path CPU_INFO_PATH = Paths.get("/proc/cpuinfo");
+
+        // cache physical processor count, so that it's not queried multiple times during tests
+        private static volatile int physicalProcessorCount = -1;
+
+        private MachineInfo() {}
+
+        public static int getAvailablePhysicalProcessorCount()
+        {
+            if (physicalProcessorCount != -1) {
+                return physicalProcessorCount;
+            }
+
+            String osArch = System.getProperty("os.arch");
+            // logical core count (including container cpu quota if there is any)
+            int availableProcessorCount = Runtime.getRuntime().availableProcessors();
+            int totalPhysicalProcessorCount;
+            if ("amd64".equals(osArch) || "x86_64".equals(osArch)) {
+                checkState(exists(CPU_INFO_PATH), "%s does not exist", CPU_INFO_PATH.toString());
```

This will fail on non-Linux operating systems. Also, we should ignore any errors reading or parsing this file and fallback to `availableProcessorCount`.